### ISSUE TITLE
Holes?

### ISSUE THREAD

**Issue by ulysses4ever** [boyland/sasylf#102]
tldr; I want to be able to write:

```
proof by lemma subst on _, p2, _
```

Some programming languages and proof systems support the concept of a hole ([Haskell](https://downloads.haskell.org/~ghc/8.10.2/docs/html/users_guide/glasgow_exts.html#typed-holes), [Idris](http://docs.idris-lang.org/en/latest/tutorial/typesfuns.html#holes), [Agda](https://williamdemeo.github.io/2014/02/27/learn-you-an-agda/#hole-filling); cf. discussion and examples under these links). It allows you to replace any part of  (almost?) any expression with a special marker  (some popular candidates: `_`, `?`, `{}`). This makes the source parseable and analyziable by the system but just not runnable. Technically this would be a compilation error. As a feedback, the system would usually present you with some information about the hole and its environment.

This would also save some time on typing “unproved", as I assume, we should be able to write `proof by _` — a moral equivalent of `proof by unproved`.

**Comment by boyland** [comment]
Looks reasonable.  "_" is the proper syntax.

**Comment by david-christiansen** [comment]
A downside of this syntax is that there's no way to put a provisional
expression in a hold and ask the system to diff their types, the way one
can in Agda.

**Comment by ulysses4ever** [comment]
Good point David! Althought, I think this is a minor issue for SASyLF, because (in my experience) it's only names of the statements that you usually put in there, unlike Agda.

**Comment by rws481664swr** [comment]
It could definitely be a good way to move ahead if you're stuck. _ would be good in place, but I think there could be something in the justification as well. It might also be desirable to get a "partial" error, or something similar to a warning.

**Comment by boyland** [comment]
_ gives an undeclared error, but quick fix can give a replacement.
